diff -up alsa-plugins-1.0.25/a52/pcm_a52.c.ffmpeg54 alsa-plugins-1.0.25/a52/pcm_a52.c
--- alsa-plugins-1.0.25/a52/pcm_a52.c.ffmpeg54	2012-01-25 08:57:07.000000000 +0100
+++ alsa-plugins-1.0.25/a52/pcm_a52.c	2012-07-11 00:03:05.155570317 +0200
@@ -444,13 +444,25 @@ static int a52_prepare(snd_pcm_ioplug_t
 #if LIBAVCODEC_VERSION_MAJOR > 52 || (LIBAVCODEC_VERSION_MAJOR == 52 && LIBAVCODEC_VERSION_MINOR >= 3)
 	switch (io->channels) {
 	case 2:
+#if LIBAVCODEC_VERSION_MAJOR < 53
 		rec->avctx->channel_layout = CH_LAYOUT_STEREO;
+#else
+		rec->avctx->channel_layout = AV_CH_LAYOUT_STEREO;
+#endif
 		break;
 	case 4:
+#if LIBAVCODEC_VERSION_MAJOR < 53
 		rec->avctx->channel_layout = CH_LAYOUT_QUAD;
+#else
+		rec->avctx->channel_layout = AV_CH_LAYOUT_QUAD;
+#endif
 		break;
 	case 6:
+#if LIBAVCODEC_VERSION_MAJOR < 53
 		rec->avctx->channel_layout = CH_LAYOUT_5POINT1;
+#else
+		rec->avctx->channel_layout = AV_CH_LAYOUT_5POINT1;
+#endif
 		break;
 	default:
 		break;
